test/src/materials/BatchTestMaterial.C

```diff
@@ -0,0 +1,62 @@
+//* This file is part of the MOOSE framework
+//* https://www.mooseframework.org
+//*
+//* All rights reserved, see COPYRIGHT for full restrictions
+//* https://github.com/idaholab/moose/blob/master/COPYRIGHT
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#include "BatchTestMaterial.h"
+
+registerMooseObject("MooseTestApp", BatchTestMaterial);
+
+InputParameters
+BatchTestMaterial::validParams()
+{
+  // we couple the same data as the BatchMaterialTest UO to compare the results of
+  // batch and regular computation
+  auto params = Material::validParams();
+  params.addCoupledVar("var1", "A coupled variable");
+  params.addRequiredParam<MaterialPropertyName>("prop1", "A RankTwoTensor property");
+  params.addRequiredParam<MaterialPropertyName>("prop2", "A Real property");
+  params.addRequiredParam<UserObjectName>("batch_uo",
+                                          "User object that performs the batch computation");
+  return params;
+}
+
+BatchTestMaterial::BatchTestMaterial(const InputParameters & parameters)
+  : Material(parameters),
+    _var1(coupledValue("var1")),
+    _prop1(getMaterialProperty<RankTwoTensor>("prop1")),
+    _prop2(getMaterialProperty<Real>("prop2")),
+    _prop_out(declareProperty<Real>("batch_out")),
+    _batch_uo(getUserObject<BatchMaterialTest>("batch_uo")),
+    _output(_batch_uo.getOutputData())
+{
+}
+
+void
+BatchTestMaterial::computeProperties()
+{
+  if (!_batch_uo.outputReady())
+    return;
```

Comment:
This is the concept that we can hopefully remove in the future right?

Reply:
I'm getting less optimistic by the line of code that I'm adding in #21804. Discrete materials are a minor design disaster. Or rather how they are being abused with custom APIs that circumvent `computePropertiesAtQp`. We should probably have a discussion about this at some point. Example: You are outputting material properties from a discrete material that is called a) through a custom API (i.e. MOOSE has no idea how to update its properties), or b) meant to be called by two different compute-true materials. a) if we do not compute the calling material the discrete material will not be updated, and there is no way of doing that because the calling material is using a custom API instead of `computePropertiesAtQp`. b) since two materials use the same discrete material, we do not know which is the state of the properties that need to be outputted.

Reply:
I agree. There could be multiple ways of improving the design of discrete material. The way I like the most is completely disallowing discrete materials to declare any material property.

Reply:
Adding to my last comment, I think of discrete materials as objects providing common APIs which are aware of the material data.

And my argument is API objects shouldn't store data...

Reply:
> completely disallowing discrete materials to declare any material property

But then you need a custom API to pass data around. By the way, we _already have_ this, it's called `DiscreteElemenUserObject` and I added that to MOOSE a few months _before_ the discrete materials idea came up.

Reply:
> > completely disallowing discrete materials to declare any material property
> 
> But then you need a custom API to pass data around. By the way, we _already have_ this, it's called `DiscreteElemenUserObject` and I added that to MOOSE a few months _before_ the discrete materials idea came up.

We don't need to pass material data around through the API, the discrete material should still be able to access material properties through the getters.

Reply:
> discrete material should still be able to access material properties through the getters.

And how doe we get the result of the discrete material calculation out again?

Reply:
The result is either returned by the method or filled into a reference. Basically the discrete material wouldn't know anything about the "state".

Reply:
Yeah, ok, then we have replicated `DiscreteElementUserObject`. Returning by the method would mean custom API for custom return types.

Reply:
We should start this discussion more formally, probably somewhere else, starting with clearly defining the scope and intended use cases of discrete materials.

Reply:
Someone create and issue and then @dschwen you can mark this conversation as resolved

Reply:
I opened #21847 to capture this